python/cuda_parallel/cuda/parallel/experimental/_iterators.py

```diff
@@ -171,274 +154,83 @@ def codegen(context, builder, sig, args):
-        return self.size
-
-
-class CountingIterator:
-    def __init__(self, count):
-        ntype = numba.from_dtype(count.dtype)
-        thisty = numba.types.CPointer(ntype)
-        self.count = _ctypes_type_given_numba_type(ntype)(count)
-        self.ntype = ntype
-        self.prefix = "count_" + ntype.name
-        self.ltoirs = [
-            _ncc(
-                f"{self.prefix}_advance",
-                CountingIterator.count_advance,
-                numba.types.void(thisty, _DISTANCE_NUMBA_TYPE),
-            ).ltoir,
-            _ncc(
-                f"{self.prefix}_dereference",
-                CountingIterator.count_dereference,
-                (thisty,),
-            ).ltoir,
-        ]
-
-    # Exclusively for numba.cuda.compile (this is not an actual method).
-    def count_advance(this, diff):
-        this[0] += diff
-
-    # Exclusively for numba.cuda.compile (this is not an actual method).
-    def count_dereference(this):
-        return this[0]
-
-    @property
-    def state_c_void_p(self):
-        return ctypes.cast(ctypes.pointer(self.count), ctypes.c_void_p)
+        cvalue = ctypes.c_void_p(ptr)
+        value_type = ntype
+        numba_type = numba.types.CPointer(numba.types.CPointer(value_type))
+        super().__init__(cvalue=cvalue, value_type=value_type, numba_type=numba_type)
 
-    @property
-    def size(self):
-        return self.ntype.bitwidth // 8
+    @staticmethod
+    def advance(it, distance):
+        it[0] = it[0] + distance
 
-    @property
-    def alignment(self):
-        return self.size
+    @staticmethod
+    def dereference(it):
+        return load_cs(it[0])
 
 
-class TransformIteratorImpl:
-    def __init__(
-        self,
-        it,
-        op_name,
-        op_return_ntype,
-        ltoirs,
-    ):
-        self.it = it
-        self.ntype = op_return_ntype
-        self.prefix = f"transform_{it.prefix}_{op_name}"
-        self.ltoirs = ltoirs
+class ConstantIterator(IteratorBase):
+    def __init__(self, value):
+        value_type = numba.from_dtype(value.dtype)
+        cvalue = _ctypes_type_given_numba_type(value_type)(value)
+        numba_type = numba.types.CPointer(value_type)
+        super().__init__(cvalue=cvalue, value_type=value_type, numba_type=numba_type)
 
-    @property
-    def state_c_void_p(self):
-        return self.it.state_c_void_p
-
-    @property
-    def size(self):
-        return self.it.size  # TODO fix for stateful op
-
-    @property
-    def alignment(self):
-        return self.it.alignment  # TODO fix for stateful op
-
-
-def TransformIterator(op, it):
-    # TODO(rwgk): Resolve issue #3064
-
-    if hasattr(it, "dtype"):
-        assert not hasattr(it, "ntype")
-        it = pointer(it, getattr(numba.types, str(it.dtype)))
-    it_ntype_ir = _ir_type_given_numba_type(it.ntype)
-    if it_ntype_ir is None:
-        raise RuntimeError(f"Unsupported: {type(it.ntype)=}")
-
-    op_abi_name = f"{op.__name__}_{it.ntype.name}"
```

Comment:
critical: I think we need a unique name per jit compilation (invokation of Parallel C `*_build` step). Some algorithms, say, reduce by key, have multiple iterators. If you try passing transform of int and transform of float, compiling member functions of corresponding iterators would result in the same symbol name: `advance` and `dereference` regardless of the operator and data type they work with. Same with counting, constant, and pointer iterators. I'd suggest to be on the safe side and have "mangled" name for everything like we used to.

Reply:
Thanks for clarifying. I worry that without some additional information, the name of the iterator and its ops + value types may not be enough: for example, what if we have 2 `ConstantIterators` with the same value type, differing only in their state?

```python
In [4]: CountingIterator(np.int32(1)).prefix
Out[4]: 'count_int32'

In [5]: CountingIterator(np.int32(2)).prefix
Out[5]: 'count_int32'
```

Should we include a unique UUID for safety?

Reply:
We don't need unique UUID here for safety. Value type on constant iterator defines codegen, but it's applied to different states. If you think about C++ analogy, there's only one `counting_iterator<int>::operator*` symbol, only `this` is different between invocations.

Reply:
We have to be careful with transform, though. `op` name is not bijective:

```python
def make_transform(constant):
  def op(val):
    return val * constant
  return iterators.Transform(..., op)

reduce_by_key(make_transform(1), make_transform(2.0), ...)
``` 

If dereference and advance are mangled by the op name alone, there's going to be a problem. So we need to have a way to distinguish operators.

Reply:
Thanks - I fixed the ABI name for TransformIterator instances. Can you check if this is good now?